File: client/src/taco/util/drive/DriveGeometry.java
package taco.util.drive;

import org.apache.commons.math3.geometry.euclidean.threed.Vector3D;

import hso.autonomy.util.geometry.Angle;
import hso.autonomy.util.geometry.IPose2D;
import hso.autonomy.util.geometry.Pose2D;
import hso.autonomy.util.misc.FuzzyCompare;

public class DriveGeometry
{
	private static final double THRESHOLD = 5;

	private static final double MINIMUM_RADIUS = 0.5;

	private static final double CAR_POSE_ADVANCE = 0.15;

	private static final double BACKWARDS_THRESHOLD = -0.55;

	/** the distance between the front and rear axle */
	private double axleSpacing;

	private Angle minAngle;

	private Angle maxAngle;

	private SteerInstruction previousInstruction;

	public DriveGeometry(double axleSpacing, Angle minAngle, Angle maxAngle)
	{
		this.axleSpacing = axleSpacing;
		this.minAngle = minAngle;
		this.maxAngle = maxAngle;
		previousInstruction = new SteerInstruction();
	}

	/**
	 * Calculates the 2D position and orientation on a circular arc with the passed length and current rotation
	 * @param length the length of the arc segment
	 * @param exitAngle the angle of the end of the arc relative to the start
	 * @return the pose of the end of the arc relative to the start
	 */
	public static Pose2D calculateArcPose(double length, Angle exitAngle)
	{
		double x = length;
		double y = 0;
		double angle = exitAngle.radians();
		if (!FuzzyCompare.isZero(angle)) {
			double radius = length / angle;
			x = Math.sin(angle) * radius;
			y = radius * (1 - Math.cos(angle));
		}

		return new Pose2D(x, y, exitAngle);
	}

	public void resetPreviousInstruction()
	{
		previousInstruction = new SteerInstruction();
	}

	public void resetPreviousInstruction(Angle angle, boolean driveForwards)
	{
		previousInstruction = new SteerInstruction(angle, driveForwards);
	}

	/**
	 * Returns a SteerInstruction-Object, which contains information about the direction that should be driven,
	 * as well as the angle.
	 * @param forceBackwards
	 */
	public SteerInstruction getNextInstruction(IPose2D currentPose, IPose2D nextWaypoint, IPose2D waypointAfterNext,
			boolean allowBackwards, boolean forceBackwards)
	{
		SteerInstruction steerInstruction =
				getSteerInstruction(currentPose, nextWaypoint, waypointAfterNext, allowBackwards, forceBackwards);
		if (steerInstruction.steeringAngle == null) {
			steerInstruction = previousInstruction;
		}

		if (steerInstruction.driveForward != previousInstruction.driveForward) {
			double distanceDotProduct = getPoseDotProduct(currentPose, nextWaypoint);

			// Prevent "flickering" of forwards and backwards-driving
			if ((distanceDotProduct < DriveGeometry.BACKWARDS_THRESHOLD + 0.2 &&
						distanceDotProduct > DriveGeometry.BACKWARDS_THRESHOLD - 0.2) ||
					currentPose.getDistanceTo(nextWaypoint) < 0.1) {
				steerInstruction = previousInstruction;
			}
		}
		previousInstruction = steerInstruction;
		return steerInstruction;
	}

	SteerInstruction getSteerInstruction(IPose2D currentPose, IPose2D nextWaypoint, IPose2D waypointAfterNext,
			boolean allowBackwards, boolean forceBackwards)
	{
		if (currentPose.getDistanceTo(nextWaypoint) < axleSpacing && waypointAfterNext != null) {
			return getSteerInstruction(currentPose, waypointAfterNext, null, allowBackwards, forceBackwards);
		}
		if (!allowBackwards) {
			return new SteerInstruction(
					getSteeringAngle(currentPose, nextWaypoint, axleSpacing, minAngle, maxAngle), true);
		} else if (forceBackwards) {
			return new SteerInstruction(
					getBackwardsSteeringAngle(currentPose, nextWaypoint, minAngle, maxAngle), false);
		}
		boolean driveBackward = shouldDriveBackwards(currentPose, nextWaypoint);
		Angle angle;
		if (!driveBackward) {
			angle = getSteeringAngle(currentPose, nextWaypoint, axleSpacing, minAngle, maxAngle);
		} else {
			angle = getBackwardsSteeringAngle(currentPose, nextWaypoint, minAngle, maxAngle);
		}

		return new SteerInstruction(angle, !driveBackward);
	}

	/**
	 * Calculate the driving pose based on the rear wheel spacing and their rolling-distances.
	 *
	 * Calculate the pose of the (center of the) rear axle after rolling leftWheelDistance and rightWheelDistance on the
	 * wheels. In order to to so, the wheel spacing (distance between the wheels along the axle).
	 *
	 * @param leftWheelDistance - the distance the left wheel progressed
	 * @param rightWheelDistance - the distance the right wheel progressed
	 * @param wheelSpacing - the distance between the two wheels along the axle
	 */
	static Pose2D calculateCurvePose(double leftWheelDistance, double rightWheelDistance, double wheelSpacing)
	{
		double x, y, alpha;

		if (leftWheelDistance == rightWheelDistance) {
			// Both wheels spin in the same direction with the same speed, thus we are driving in a straight line.
			x = rightWheelDistance;
			y = 0;
			alpha = 0;

		} else if (leftWheelDistance == -rightWheelDistance) {
			// Both wheels spin at the same speed, but in different directions, thus we are turing on spot.
			x = 0;
			y = 0;
			alpha = 2 * rightWheelDistance / wheelSpacing;

		} else {
			double innerWheelDistance = leftWheelDistance;
			double outerWheelDistance = rightWheelDistance;
			int sideFactor = 1;

			if (Math.abs(leftWheelDistance) > Math.abs(rightWheelDistance)) {
				innerWheelDistance = rightWheelDistance;
				outerWheelDistance = leftWheelDistance;
				sideFactor = -1;
			}

			if (leftWheelDistance * rightWheelDistance >= 0) {
				// Both wheels spin in the same direction, thus the origin of the driving circle is outside (left or
				// right) the car.

				double diffDistance = outerWheelDistance - innerWheelDistance;
				double r = (wheelSpacing * innerWheelDistance / diffDistance) + (wheelSpacing / 2);

				alpha = sideFactor * (diffDistance / wheelSpacing);
				x = sideFactor * (Math.sin(alpha) * r);
				y = -sideFactor * (Math.cos(alpha) * r - r);

			} else {
				// The wheels spin in different directions, thus the origin of the driving circle is between the two
				// wheels.

				double distanceSum = innerWheelDistance + outerWheelDistance;
				double r = (wheelSpacing / 2) - (wheelSpacing * innerWheelDistance / distanceSum);

				alpha = sideFactor * (distanceSum / wheelSpacing);
				x = sideFactor * (Math.sin(alpha) * r);
				y = -sideFactor * (Math.cos(alpha) * r - r);
			}
		}

		return new Pose2D(x, y, Angle.rad(alpha));
	}

	/**
	 * Calculate the curve radius from a certain steering angle.
	 *
	 * Calculate the resulting curve radius by driving with the given steering angle.
	 * Note: A positive/negative steering angle results in an positive/negative curve radius.
	 *
	 * @param steeringAngle - the steering angle
	 * @return the curve radius resulting from the given steering angle
	 */
	public double calculateCurveRadius(Angle steeringAngle)
	{
		double steering = steeringAngle.radians();
		if (Math.abs(steering) >= Math.PI / 2) {
			return 0;
		} else if (steering == 0) {
			return Double.MAX_VALUE;
		}

		return axleSpacing / Math.tan(steering);
	}

	/**
	 * Calculate the steering angle to drive a certain curve.
	 *
	 * Calculate the necessary steering angle to drive a curve with the given radius.
	 * Note: A positive/negative curve radius results in an positive/negative steering angle.
	 *
	 * @param curveRadius - the radius of the intended curve in meter
	 * @return the steering angle resulting in the given curve radius
	 */
	Angle calculateSteeringAngle(double curveRadius)
	{
		if (curveRadius < 0.000001 && curveRadius >= 0.0) {
			return Angle.ANGLE_90;
		} else if (curveRadius > -0.000001 && curveRadius <= 0.0) {
			return Angle.ANGLE_90.negate();
		}

		return Angle.rad(Math.atan(Math.abs(axleSpacing) / curveRadius));
	}

	public Circle2D getCircle(IPose2D currentPose, IPose2D nextWaypoint, IPose2D backAxlePose, boolean allowBackwards)
	{
		Circle2D circle = new Circle2D(nextWaypoint, currentPose.getPosition());

		if (Double.isNaN(circle.getRadius())) {
			return null;
		} else if (circle.getDirectionOfRotation(nextWaypoint) != circle.getDirectionOfRotation(currentPose)) {
			if (backAxlePose != null) {
				return getCircle(backAxlePose, nextWaypoint, null, allowBackwards);
			}
			if (!allowBackwards) {
				return null;
			}
		} else if (circle.getRadius() < MINIMUM_RADIUS) {
			//			return null;
		}
		return circle;
	}

	public Line2D getLine(IPose2D currentPose, IPose2D nextWaypoint, boolean allowBackwards)
	{
		Line2D line = new Line2D(currentPose.getPosition(), nextWaypoint.getPosition());
		if (!allowBackwards && Math.abs(line.getAngle().subtract(currentPose.getAngle()).degrees()) > THRESHOLD) {
			return null;
		} else if (allowBackwards &&
				   Math.abs(line.getAngle().subtract(nextWaypoint.getAngle()).degrees()) > THRESHOLD) {
			return null;
		}
		return line;
	}

	IBehaviorGeometry chooseDriveGeometry(
			IPose2D currentPose, IPose2D nextWaypoint, IPose2D backAxlePose, boolean allowBackwards)
	{
		IBehaviorGeometry behaviorGeometry = getLine(currentPose, nextWaypoint, allowBackwards);
		if (behaviorGeometry != null) {
			return behaviorGeometry;
		}
		behaviorGeometry = getCircle(currentPose, nextWaypoint, backAxlePose, allowBackwards);
		if (behaviorGeometry != null) {
			return behaviorGeometry;
		}
		return null;
	}

	/**
	 * Returns an angle that can be used to reach a pose from another.
	 * @param currentPose - an initial, current Pose
	 * @param nextWaypoint - the pose that shall be reached starting from currentPose
	 * @return Angle An Angle that can be driven in order to get closer to nextWaypoint
	 */
	private Angle getSteeringAngle(
			IPose2D currentPose, IPose2D nextWaypoint, double axleSpacing, Angle minAngle, Angle maxAngle)
	{
		IPose2D backAxlePose = currentPose;
		currentPose = currentPose.applyTo(new Pose2D(axleSpacing, 0));

		IBehaviorGeometry geometry = chooseDriveGeometry(currentPose, nextWaypoint, backAxlePose, false);
		if (geometry == null) {
			return null;
		}
		Vector3D virtualPoint = currentPose.getPosition().add(
				AngleUtils.get3DVectorFromAngle(currentPose.getAngle().radians(), CAR_POSE_ADVANCE));

		IPose2D globalVirtualPose = geometry.getClosestPose(virtualPoint);

		IPose2D target = currentPose.applyInverseTo(globalVirtualPose);

		Angle angle = AngleUtils.to(target.getX(), target.getY());
		return AngleUtils.limit(angle, minAngle, maxAngle);
	}

	private Angle getBackwardsSteeringAngle(IPose2D currentPose, IPose2D backwardsPoint, Angle minAngle, Angle maxAngle)
	{
		IBehaviorGeometry geometry = chooseDriveGeometry(backwardsPoint, currentPose, null, true);
		if (geometry == null) {
			return null;
		}
		Angle backwardsAngle = currentPose.getAngle();

		if (geometry.getClass().getName().equals(Line2D.class.getName())) {
			backwardsAngle = ((Line2D) geometry).getAngle();
		}
		Vector3D virtualPoint =
				currentPose.getPosition().subtract(AngleUtils.get3DVectorFromAngle(backwardsAngle.radians(), 0.15));
		IPose2D globalVirtualPose;

		globalVirtualPose = geometry.getClosestPose(virtualPoint);
		IPose2D target;
		// This distinction is important for accuracy:
		if (geometry.getClass().getName().equals(Circle2D.class.getName())) {
			target = currentPose.applyInverseTo(globalVirtualPose);
		} else {
			target = globalVirtualPose.applyInverseTo(currentPose);
		}

		Angle angle = AngleUtils.to(target.getX(), target.getY());
		return AngleUtils.limit(angle, minAngle, maxAngle);
	}

	double getPoseDotProduct(IPose2D currentPose, IPose2D nextPose)
	{
		Line2D frontAxleLine = new Line2D(currentPose.getX(), currentPose.getY(), currentPose.getAngle(), axleSpacing);

		Vector3D subtract = new Vector3D(frontAxleLine.getEnd().getX(), frontAxleLine.getEnd().getY(), 0)
									.subtract(currentPose.getPosition());
		if (subtract.getNorm() < 0.0001) {
			return 1;
		}
		Vector3D carVector = subtract.normalize();

		Vector3D waypointVector = nextPose.getPosition().subtract(currentPose.getPosition());
		if (waypointVector.getNorm() < 0.0001) {
			return 1;
		}
		waypointVector = waypointVector.normalize();
		return waypointVector.dotProduct(carVector);
	}

	public static double getPoseDotProduct(IPose2D currentPose, IPose2D nextPose, double axleSpacing)
	{
		Line2D frontAxleLine = new Line2D(currentPose.getX(), currentPose.getY(), currentPose.getAngle(), axleSpacing);

		Vector3D subtract = new Vector3D(frontAxleLine.getEnd().getX(), frontAxleLine.getEnd().getY(), 0)
									.subtract(currentPose.getPosition());
		if (subtract.getNorm() < 0.0001) {
			return 1;
		}
		Vector3D carVector = subtract.normalize();

		Vector3D waypointVector = nextPose.getPosition().subtract(currentPose.getPosition());
		if (waypointVector.getNorm() < 0.0001) {
			return 1;
		}
		waypointVector = waypointVector.normalize();
		return waypointVector.dotProduct(carVector);
	}

	private boolean shouldDriveBackwards(IPose2D currentPose, IPose2D nextPose)
	{
		return getPoseDotProduct(currentPose, nextPose) < BACKWARDS_THRESHOLD;
	}

	//	public IBehaviorGeometry getSCurveGeometry(
	//			IPose2D currentPose, IPose2D nextWaypoint, IPose2D startPoint, SegmentType currentSegmentType)
	//	{
	//		if (currentSegmentType == SegmentType.S_CURVE_BOTTOM ||
	//				currentSegmentType ==
	//						SegmentType.S_CURVE_TOP) { // We most likely have a S Curve - as long as we drive via
	//												   // map with them marked. It'd be better to implement multiple
	//												   // conditions though, since this isn't sustainable.
	//			return new SCurve(startPoint, nextWaypoint);
	//		}
	//		return null;
	//	}
}

File: server/src/hsogserver/utils/geometry/drive/DriveGeometry.cpp
#include "DriveGeometry.h"
#include "utils/geometry/Geometry.h"

#include <cmath>
#include <limits>

using namespace taco;
using namespace Eigen;

#define THRESHOLD 5
#define MINIMUM_RADIUS 0.5

const Pose2D DriveGeometry::calculateCurvePose(
		const double &leftWheelDistance, const double &rightWheelDistance, const double &wheelSpacing)
{
	double x, y, alpha;

	if (leftWheelDistance == rightWheelDistance) {
		// Both wheels spin in the same direction with the same speed, thus we are driving in a straight line.
		x = 0;
		y = leftWheelDistance;
		alpha = 0;
	} else if (leftWheelDistance == -rightWheelDistance) {
		// Both wheels spin at the same speed, but in different directions, thus we are turing on spot.
		x = 0;
		y = 0;
		alpha = 2 * leftWheelDistance / wheelSpacing;
	} else {
		double innerWheelDistance = leftWheelDistance;
		double outerWheelDistance = rightWheelDistance;
		int sideFactor = 1;

		if (std::abs(leftWheelDistance) > std::abs(rightWheelDistance)) {
			innerWheelDistance = rightWheelDistance;
			outerWheelDistance = leftWheelDistance;
			sideFactor = -1;
		}

		if (leftWheelDistance * rightWheelDistance >= 0) {
			// Both wheels spin in the same direction, thus the origin of the driving circle is outside (left or right)
			// the car.

			double diffDistance = outerWheelDistance - innerWheelDistance;
			double r = (wheelSpacing * innerWheelDistance / diffDistance) + (wheelSpacing / 2);

			alpha = sideFactor * (diffDistance / wheelSpacing);
			x = sideFactor * (std::cos(alpha) * r - r);
			y = sideFactor * (std::sin(alpha) * r);
		} else {
			// The wheels spin in different directions, thus the origin of the driving circle is between the two wheels.

			double distanceSum = innerWheelDistance + outerWheelDistance;
			double r = (wheelSpacing / 2) - (wheelSpacing * innerWheelDistance / distanceSum);

			alpha = sideFactor * (distanceSum / wheelSpacing);
			x = sideFactor * (std::cos(alpha) * r - r);
			y = sideFactor * (std::sin(alpha) * r);
		}
	}

	return Pose2D(x, y, Angle::rad(alpha));
}

const double DriveGeometry::calculateCurveRadius(const Angle &steeringAngle, const double &axleSpacing)
{
	if (std::abs(steeringAngle.rad()) >= M_PI / 2) {
		return 0;
	} else if (steeringAngle.rad() == 0) {
		return std::numeric_limits<double>::max();
	}

	return axleSpacing / std::tan(steeringAngle.rad());
}

const Angle DriveGeometry::calculateSteeringAngle(const double &curveRadius, const double &axleSpacing)
{
	if (curveRadius == 0) {
		return Angle::Deg_90();
	} else if (curveRadius == -0) {
		return Angle::Deg_N90();
	}

	return Angle::rad(std::atan(std::abs(axleSpacing) / curveRadius));
}

const Driveable DriveGeometry::getLineBeforeCircleGeometry(const Pose2D &carPose, const Pose2D &nextWayPoint)
{
	Pose2D nextWayPointTransformed = carPose / nextWayPoint;
	if ((nextWayPointTransformed.y() > 0 && nextWayPointTransformed.getAngle().rad() < 0) ||
			(nextWayPointTransformed.y() < 0 && nextWayPointTransformed.getAngle().rad() > 0))
		return Driveable::Invalid();

	double x = Geometry::getTouchingPoint(nextWayPointTransformed);
	// if first point on circle is behind car pose -> circle starts behind car pose
	if (x < 0)
		return Driveable::Invalid();

	Vector2d lineEnd = carPose * Vector2d(x, 0);
	Pose2D intercept(lineEnd, carPose.getAngle());
	Circle2D circle(nextWayPoint, lineEnd);
	// if circle origin is behind nextWayPoint, path is in wrong direction
	if ((nextWayPoint / circle.origin())(0) < 0)
		return Driveable::Invalid();
	// if intercept is behind car pose
	//   else if((carPose / intercept).x() > 0)
	//     return Driveable::Invalid();
	// if angle is same
	else if (fabs(carPose.getAngle().rad() - nextWayPoint.getAngle().rad()) < 0.0001)
		return Driveable::Invalid();
	// if line points in wrong direction: )_ instead of _)
	else if (circle.getDirectionOfRotation(nextWayPoint) != circle.getDirectionOfRotation(intercept))
		return Driveable::Invalid();
	// if radius is too small
	else if (circle.radius() < MINIMUM_RADIUS)
		return Driveable::Invalid();
	else
		return Driveable(IManeuverGeometry::Ptr(new Line2D(carPose.getPosition(), lineEnd)),
				IManeuverGeometry::Ptr(new Circle2D(circle)), intercept);
}

const Driveable DriveGeometry::getCircleBeforeLineGeometry(const Pose2D &carPose, const Pose2D &nextWayPoint)
{
	Pose2D carPoseTransformed = nextWayPoint / carPose;
	if ((carPoseTransformed.y() > 0 && carPoseTransformed.getAngle().rad() < 0) ||
			(carPoseTransformed.y() < 0 && carPoseTransformed.getAngle().rad() > 0))
		return Driveable::Invalid();

	double x = Geometry::getTouchingPoint(carPoseTransformed);
	// if point is in front of next way point -> circle ends behind nextWayPoint
	if (x > 0)
		return Driveable::Invalid();

	Vector2d lineStart = nextWayPoint * Vector2d(x, 0);
	Circle2D circle(carPose, lineStart);
	Pose2D intercept(lineStart, nextWayPoint.getAngle());
	// if circle origin is behind carPose, path is in wrong direction
	if ((carPose / circle.origin())(0) < 0)
		return Driveable::Invalid();
	// if angle is same
	else if (fabs(carPose.getAngle().rad() - nextWayPoint.getAngle().rad()) < 0.0001)
		return Driveable::Invalid();
	// if intercept is behind next way point
	//   else if((nextWayPoint / intercept).x() < 0)
	//     return Driveable::Invalid();
	// if line points in wrong direction: )_ instead of _)
	else if (circle.getDirectionOfRotation(intercept) != circle.getDirectionOfRotation(carPose))
		return Driveable::Invalid();
	// if radius is too small
	else if (circle.radius() < MINIMUM_RADIUS)
		return Driveable::Invalid();
	else
		return Driveable(IManeuverGeometry::Ptr(new Circle2D(circle)),
				IManeuverGeometry::Ptr(new Line2D(lineStart, nextWayPoint.getPosition())), intercept);
}

const Driveable DriveGeometry::getCircle(const Pose2D &carPose, const Pose2D &nextWayPoint)
{
	Circle2D circle(nextWayPoint, carPose.getPosition());

	if (fabs((circle.slope(carPose.getPosition()) - carPose.getAngle()).deg()) > THRESHOLD &&
			fabs((circle.slope(carPose.getPosition()).opposite() - carPose.getAngle()).deg()) > THRESHOLD)
		return Driveable::Invalid();
	else if (circle.getDirectionOfRotation(nextWayPoint) != circle.getDirectionOfRotation(carPose))
		return Driveable::Invalid();
	else if (circle.radius() < MINIMUM_RADIUS)
		return Driveable::Invalid();
	else
		return Driveable(IManeuverGeometry::Ptr(new Circle2D(circle)));
}

const Driveable DriveGeometry::DriveGeometry::getLine(const Pose2D &carPose, const Pose2D &nextWayPoint)
{
	Line2D line(carPose.getPosition(), nextWayPoint.getPosition());
	// if differences of angle from carPose and nextWayPoint is higher than threshold
	if (fabs((line.getAngle() - carPose.getAngle()).deg()) > THRESHOLD ||
			fabs((line.getAngle() - nextWayPoint.getAngle()).deg()) > THRESHOLD)
		return Driveable::Invalid();
	else
		return Driveable(IManeuverGeometry::Ptr(new Line2D(line)));
}

const Driveable DriveGeometry::getSCurveBeforeLineGeometry(const Pose2D &carPose, const Pose2D &nextWayPoint)
{
	double angleDiff = fabs((carPose.getAngle() - nextWayPoint.getAngle()).deg());
	// catch threshold
	if (angleDiff > 45 && angleDiff < 135)
		return Driveable::Invalid();

	Cosine geometry = getConnectingCosine(nextWayPoint, carPose);
	geometry.normalizeRatioToPoint(carPose.getPosition());
	Pose2D nextWayPointTransformed = geometry.origin() / nextWayPoint;
	// check if curve is too long for next way point
	if (nextWayPointTransformed.x() < 0)
		return Driveable::Invalid();
	// high elongation ^= high steering angle
	else if (geometry.elongation() > 14.0)
		return Driveable::Invalid();

	// line starts right or left of origin of cosine
	Vector2d lineStart = geometry.origin() * Vector2d(0, geometry.spreading());
	Line2D line(lineStart, nextWayPoint.getPosition());

	return Driveable(IManeuverGeometry::Ptr(new Cosine(geometry)), IManeuverGeometry::Ptr(new Line2D(line)),
			Pose2D(lineStart, nextWayPoint.getAngle()));
}

const Cosine DriveGeometry::getConnectingCosine(const Pose2D &directionLocal, const Pose2D &other)
{
	Pose2D otherTransformed = directionLocal / other;

	double slope = tan(otherTransformed.getAngle().rad());
	bool directionChanged = false;
	if (otherTransformed.y() > 0) {
		directionChanged = true;
		slope *= -1;
	}
	double x = asin(slope) + M_PI;
	double yDiff = fabs(otherTransformed.y());
	double yOffset = fabs(cos(x)) + 1;
	double spreading = yDiff / yOffset;
	if (directionChanged)
		spreading *= -1;

	double periodLength = fabs((2.0 * M_PI * otherTransformed.x()) / x);
	double elongation = (2 * M_PI) / periodLength;

	Pose2D originTransformed = Pose2D(0, -spreading, Angle());
	Pose2D origin = directionLocal * originTransformed;
	return Cosine(origin, elongation, spreading);
}
